DAOS-16916 container: use ABT mutex to serialize cont open

wangdi1 opened the pull request:
Use ABT mutex to serialize cont open.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

wangdi1 committed:
DAOS-16916 container: use ABT mutex to serialize cont open

Use ABT mutex to serialize cont open.

Signed-off-by: Di Wang <ddiwang@google.com>

jolivier23 commented on src/include/daos_srv/container.h:
```diff
@@ -71,7 +71,7 @@ struct ds_cont_child {
 	    sc_destroying : 1, sc_vos_agg_active : 1, sc_ec_agg_active : 1,
 	    /* flag of CONT_CAPA_READ_DATA/_WRITE_DATA disabled */
-	    sc_rw_disabled : 1, sc_scrubbing : 1, sc_rebuilding : 1, sc_open_initializing : 1;
+	    sc_rw_disabled : 1, sc_scrubbing : 1, sc_rebuilding : 1;
```
mentioned on your last patch but if you change these to semicolons, clang-format will look better and it's the same behavior.

wangdi1 replied:
ah, sure, missed that. Thanks!

wangdi1 committed:
Merge branch 'master' into wangdi/daos_16916_new

Signed-off-by: Di Wang <ddiwang@google.com>

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/3/execution/node/371/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/3/execution/node/375/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/3/execution/node/363/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/3/execution/node/358/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/3/execution/node/359/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/3/execution/node/327/log

wangdi1 committed:
Merge branch 'master' into wangdi/daos_16916_new

Run-GHA: true
Allow-unstable-test: true
Required-githooks: true

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/4/execution/node/360/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/4/execution/node/375/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/4/execution/node/350/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/4/execution/node/334/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/4/execution/node/354/log

wangdi1 committed:
DAOS-16916 contatiner: remove a few unnecessary open count assertion

Since cont open process might be yield due if IV fetch needs to fetch
the property remotely, so sc_open counter might be larger than 1 if
property IV fetch failed.

Run-GHA: true
Allow-unstable-test: true
Required-githooks: true
Signed-off-by: Di Wang <ddiwang@google.com>

wangdi1 commented:
@Nasf-Fan it looks like those sc_open counter might be larger than 1 if iv fetch failed? so I removed those assertion? thoughts?

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/5/execution/node/376/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/5/execution/node/327/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/5/execution/node/328/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/5/execution/node/359/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15742/5/execution/node/350/log

wangdi1 committed:
DAOS-16916 container: fix build failure

fix build failure

Run-GHA: true
Allow-unstable-test: true
Signed-off-by: Di Wang <ddiwang@google.com>

NiuYawei commented on src/container/srv_target.c:
```diff
@@ -1630,59 +1630,55 @@ ds_cont_local_open(uuid_t pool_uuid, uuid_t cont_hdl_uuid, uuid_t cont_uuid,
 		D_ASSERT(hdl->sch_cont != NULL);
 		D_ASSERT(hdl->sch_cont->sc_pool != NULL);
-
 		hdl->sch_cont->sc_open++;
```
Suggestion: simply moving ABT_mutex_lock() before sc_open++ would make code simpler, and you don't have to change the assertions in DTX code.

Nasf-Fan replied:
> Suggestion: simply moving ABT_mutex_lock() before sc_open++ would make code simpler, and you don't have to change the assertions in DTX code.

Yes, I would suggest for the same. On the other hand, what will happen if another open arrives with the same uuid during former one yield? The latter one will return directly since it can find the "hdl" in cache but without considering whether the container is really opened or in-opening. That seems incorrect.

NiuYawei replied:
@wangdi1 Could you update the PR to address this comment? Thanks!

NiuYawei reviewed (commented):
There is a conflict to be resolved.

wangdi1 committed:
Merge branch 'master' into wangdi/daos_16916_new

Signed-off-by: Di Wang <ddiwang@google.com>

wangdi1 committed:
Merge branch 'master' into wangdi/daos_16916_new

NiuYawei commented:
@wangdi1 FYI, I addressed this issue along with other fixes at https://github.com/daos-stack/daos/pull/16793 . Thanks.
